### DISCUSSION TITLE
Content-Length not set when using BodyJSON

### DISCUSSION THREAD

**Discussion by frankfil** [earthboundkid/requests#49] [Support] [answered]
I have an API that I am making a POST call to with some data like the following code snippet:

```Go
// "object" is a struct
err := requests.URL("https://example.com/api/something").BodyJSON(object).Fetch(context.Background())
````

But the actual HTTP POST is missing the Content-Length Header, causing the server to respond with a 400 Bad Request (actually it probably should be 411 Length Required but that's a discussion I've yet to have with those devs). 

The docs for [NewRequestWithContext](https://pkg.go.dev/net/http#NewRequestWithContext) say the following:

> If body is of type *bytes.Buffer, *bytes.Reader, or *strings.Reader, the returned request's ContentLength is set to its exact value (instead of -1), GetBody is populated (so 307 and 308 redirects can replay the body), and Body is set to NoBody if the ContentLength is 0.

A brief look at the `BodyJSON` func and it's returning an `io.NopCloser` so would I be correct in thinking this is why the Content-Length header is not being set?

So how can I get the header added with the correct value?

**Comment by earthboundkid** [comment, marked as answer]
The standard library is inconsistent about when it wants a Reader and when it wants a ReadCloser. I am adapting Readers into ReadClosers with io.NopCloser, but it looks like that is interfering with the Content Length setter. I think the fix is that I should write my own NopCloser type and just strip it off the body passed into NewRequestWithContext, but I’m not sure. I’ll poke around and see what I can figure out.

**Reply by earthboundkid** [reply]
See https://github.com/carlmjohnson/requests/pull/42/commits/184aed06deec02bb2f144f587828cd4a7fdf6552. I'll take another look this afternoon to see if it can be merged in.

**Reply by earthboundkid** [reply]
I think https://github.com/carlmjohnson/requests/commit/184aed06deec02bb2f144f587828cd4a7fdf6552 should definitely go in, but another way to add a content length is with a custom transport:

```go
func BufferedBodyTransport(rt http.RoundTripper) Transport {
	return RoundTripFunc(func(req *http.Request) (res *http.Response, err error) {
		var buf bytes.Buffer
		if _, err = io.Copy(&buf, req.Body); err != nil {
			return
		}
		// shallow copy req before modifying
		req2 := *req
		data := buf.Bytes()
		req2.Body = io.NopCloser(bytes.NewReader(data))
		req2.ContentLength = int64(len(data))
		req2.GetBody = func() (io.ReadCloser, error) {
			return io.NopCloser(bytes.NewReader(data)), nil
		}

		return rt.RoundTrip(&req2)
	})
}
```

Does that seem to have any other uses, or should people who need buffered bodies just save their data to a slice and use .BodyBytes?

**Reply by frankfil** [reply]
Your commit https://github.com/carlmjohnson/requests/commit/184aed06deec02bb2f144f587828cd4a7fdf6552 definitely solves the issue - thank you.

I can't think of a use for `BufferedBodyTransport` off the top of my head but others may find it useful. Even having this code to refer to as one method of modifying the transport is helpful.

**Reply by earthboundkid** [reply]
I made a wiki page and put it onto there: https://github.com/carlmjohnson/requests/wiki

**Reply by frankfil** [reply]
Tested your newly released tag and works perfectly - thank you!
